api/models/func.go

```diff
@@ -0,0 +1,224 @@
+package models
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/fnproject/fn/api/id"
+	"github.com/go-openapi/strfmt"
+)
+
+var (
+	// these are vars so that they can be configured. these apply
+	// across function & trigger (resource config)
+
+	MaxMemory      uint64 = 8 * 1024 // 8GB
+	MaxTimeout     int32  = 300      // 5m
+	MaxIdleTimeout int32  = 3600     // 1h
+)
+
+// FuncWrapper makes jason purrty for a Func.
+type FuncWrapper struct {
+	Func *Func `json:"func"`
+}
+
+// Func contains information about a function configuration.
+type Func struct {
+	// ID is the generated resource id.
+	ID string `json:"id" db:"id"`
+	// Name is a user provided name for this func.
+	Name string `json:"name" db:"name"`
+	// Image is the fully qualified container registry address to execute.
+	// examples: hub.docker.io/me/myfunc, me/myfunc, me/func:0.0.1
+	Image string `json:"image" db:"image"`
+	// ResourceConfig specifies resource constraints.
+	ResourceConfig // embed (TODO or not?)
+	// Config is the configuration passed to a function at execution time.
+	Config Config `json:"config" db:"config"`
+	// Annotations allow additional configuration of a function, these are not passed to the function.
+	Annotations Annotations `json:"annotations,omitempty" db:"annotations"`
+	// CreatedAt is the UTC timestamp when this function was created.
+	CreatedAt strfmt.DateTime `json:"created_at,omitempty" db:"created_at"`
+	// UpdatedAt is the UTC timestamp of the last time this func was modified.
+	UpdatedAt strfmt.DateTime `json:"updated_at,omitempty" db:"updated_at"`
+
+	// TODO wish to kill but not yet ?
+	// Format is the container protocol the function will accept,
+	// may be one of: json | http | cloudevent | default
+	Format string `json:"format" db:"format"`
+}
+
+// ResourceConfig specified resource constraints imposed on a function execution.
+type ResourceConfig struct {
+	// Memory is the amount of memory allotted, in MB.
+	Memory uint64 `json:"memory,omitempty" db:"memory"`
+	// CPUs is the max usable CPU cores for this route. Value in MilliCPUs,
+	// (eg. 500m) or as floating-point (eg. 0.5)
+	// TODO this is a good chance to fix this. let's pick one?
+	CPUs MilliCPUs `json:"cpus,omitempty" db:"cpus"`
+	// Timeout is the max execution time for a function, in seconds.
+	// TODO this should probably be milliseconds?
+	Timeout int32 `json:"timeout,omitempty" db:"timeout"`
+	// IdleTimeout is the
+	// TODO this should probably be milliseconds
+	IdleTimeout int32 `json:"idle_timeout,omitempty" db:"idle_timeout"`
+}
+
+// SetDefaults sets zeroed field to defaults.
+func (f *Func) SetDefaults() {
+	if f.ID == "" {
+		f.ID = id.New().String()
+	}
+
+	if f.Memory == 0 {
```

Comment:
Aren't these all natural go zero values? Only ID and time seem like they need to be here.

Reply:
we call this on the func that we deserialized from the request body, so e.g. memory could be set to 42 and we don't want to override it if that's the case